### ISSUE TITLE
Issue with hashmap algorithm and probably more when loading certain models.

### ISSUE THREAD

**Issue by BaconOFBurger** [syoyo/tinyobjloader-c#43]
See issue https://github.com/raysan5/raylib/issues/1374 for more detailed information.

**Comment by chriscamacho** [comment]
there is also another related issue here...
when looking up the hash (to assign materials to meshes) the usemtl name gets truncated (not sure why) so while there might be a "blueMat" material it attempts to look for "blueMa"
that said their is another issue still as instead of all -1's for my materials id's I'm getting all 0's - but at least the hash map look up is now working (64bit Linux)

**Comment by chriscamacho** [comment]
actually the issue is
    material_name_null_term[commands[i].material_name_len - 1] = 0;
should be
    material_name_null_term[commands[i].material_name_len] = 0;

**Comment by syoyo** [comment]
hashmap implementation is contributed by the contributor and don't know how it is implemented for me.

@BaconOFBurger Could you please post a PR to fix hashmap or rewritten hashmap implementation?(and also we may better to add a ifdef switch to use hashmap and simple linear search)

**Comment by syoyo** [comment]
> actually the issue is
> material_name_null_term[commands[i].material_name_len - 1] = 0;

such code does not exit in `main` branch: https://github.com/syoyo/tinyobjloader-c/blob/1ee072e2f282789906d5aa9e8557f22c7711d80a/tinyobj_loader_c.h#L1407

**Comment by chriscamacho** [comment]
GOOD!

I'm working on whatever version is included currently with the dev branch of raylib...

I've actually managed to get it to break out all the material meshes (in raylib) now so nearly working fully with materials!

**Comment by chriscamacho** [comment]
BTW I presume this hashmap implementation works... https://github.com/jpcy/objzero/blob/master/objzero.c - I didn't manage to trigger the issue where it can't insert - it ended up being the already fixed truncated lookup string bug

**Comment by BaconOFBurger** [comment]
> hashmap implementation is contributed by the contributor and don't know how it was implemented for me.

> @BaconOFBurger Could you please post a PR to fix hashmap or rewritten hashmap implementation?(and also we may better to add a ifdef switch to use hashmap and simple linear search)

I'm afraid I don't have the knowledge to do so, originally I found this bug accidentally loading hundreds of models for my project while few of them refused to load, after a lot of failures and lost brain cells trying to find the cause of this issue, I gave up and opened an issue in the raylib repo, this is the part where @masterex1000, @rhf104 and @chriscamacho discovered that this was a problem with the hashmap algorithm and so I forwarded that information by oppening an issue here.

TLDR; What I mean I'm unable to help more than just providing the resources required to reproduce the bug.

**Comment by syoyo** [comment]
@BaconOFBurger I have updated tinyobjloader-c(not related to hashtable. Changed file callback API) https://github.com/syoyo/tinyobjloader-c/commit/50293e537ff2842b573fe30fd2ed096f9f9e408f , and ambulance.obj seems correctly loaded with `example/viewer`, so please use the latest one.

 If you still encounter the issue with the latest code, please file an reproducible .obj/.mtl files with codes(e.g. modify `examples/viewer`)
